Behavior change of fail2ban sensor

grea09 opened the pull request:
Rebase of #17550 because of branch switching problems.

In order to fix a bug in the component to make the component meet the PR (scan_interval having no default value), I tried to solve several of the problems of the sensor.

I propose a behaviour change that expose the number of current bans for the value of the sensor since it is the kind of information often presented in reportings. I also lifted the limit of the total ban list from 10 to 1000 and fixed the regex issues with IPv6.

## Description:


**Related issue (if applicable):** fixes #10500

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#6895

## Example entry for `configuration.yaml` (if applicable):
```yaml
  - platform: fail2ban
    jails:
      - apache-auth
      - ha
      - pam-generic
      - sshd
      - webmin-auth
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

grea09 committed:
Bug fix plus behaviour change of fail2ban sensor (redo)

Rebase of #17550 because of branch switching problems.

grea09 committed:
Updating tests for fail2ban

grea09 committed:
Hound fixing

grea09 committed:
Hound fixing 2

grea09 committed:
Timedelta moved to __init__ of BanLogParser

grea09 committed:
Adding units for enabling graph view of data

grea09 committed:
Timedelta doesn't need to be imported now

grea09 committed:
Hound fixing 3

MartinHjelmare commented on homeassistant/components/sensor/fail2ban.py:
```diff
@@ -43,7 +45,7 @@
     name = config.get(CONF_NAME)
     jails = config.get(CONF_JAILS)
-    scan_interval = config.get(CONF_SCAN_INTERVAL)
+    scan_interval = config.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
```
This is not the correct approach. Please see comment here:
https://github.com/home-assistant/home-assistant/pull/10500#discussion_r150186077

grea09 replied:
I am not sure of how to do but I will try something.

grea09 replied:
I will let you check the changes, but it works in my install.

grea09 committed:
Using async update instead of internal logic

grea09 committed:
Removing update interval alltogether

grea09 committed:
updating method for updates

grea09 committed:
Last sensor.state assert to change

grea09 committed:
Trying to add self.hass.block_till_done()

grea09 commented:
I don't seem to be able to wait for async_update to finish in the test… Maybe someone knowing the project more can complete my work on that or indicate me where to look for a way to do that correctly.

MartinHjelmare commented on homeassistant/components/sensor/fail2ban.py:
```diff
@@ -81,61 +82,58 @@ def state_attributes(self):
         """Update the list of banned ips."""
-        if self.log_parser.timer():
-            self.log_parser.read_log(self.jail)
+        self.log_parser.read_log(self.jail)
```
We're not allowed to do I/O directly in coroutines, ie in the event loop. Is there any reason why we should change to `async_update`?

grea09 replied:
I obviously misunderstood your previous comment. I changed to async_update to remove the time logic from the sensor. Is there something I don't understand ?

MartinHjelmare replied:
Just call the method `update`. This will work with our standard auto polling and also work with sync I/O.

grea09 replied:
Doing so now.

MartinHjelmare replied:
Make it a regular function, ie just `def`, not `async def`.

grea09 replied:
It's done now

MartinHjelmare commented on homeassistant/components/sensor/fail2ban.py:
```diff
@@ -61,13 +58,17 @@ def __init__(self, name, jail, log_parser):
+        _LOGGER.info("Setting up jail %s", self.jail)
+        _LOGGER.debug("Regex: %s",
+                      str(self.log_parser.ip_regex[self.jail]))
+        self.log_parser.read_log(self.jail)
```
Update is called when adding the entity so reading the log here is not needed.

grea09 replied:
I needed that with the previous time logic. I will remove that.

MartinHjelmare commented on homeassistant/components/sensor/fail2ban.py:
```diff
@@ -26,8 +23,9 @@
 
 DEFAULT_NAME = 'fail2ban'
 DEFAULT_LOG = '/var/log/fail2ban.log'
-SCAN_INTERVAL = timedelta(seconds=120)
```
Keep the `SCAN_INTERVAL` constant defined. This will set the default scan interval for the platform.

MartinHjelmare reviewed (commented):
I think we should split this PR in two PRs. One (this) that fixes the bug with the scan interval. The other should propose the change of behavior. The latter will be a breaking change, so it might not be accepted.

grea09 commented:
Well the previous limit was 10 on the total_ban attribute but no limit on current_ban. I too don't understand the need for this attribute in the first place. I raised the limit because the only reason for the total_ban attribute to be useful is to contain more than the current_ban list.

I think that a discussion is needed on this component that really seemed to have been left out for a while. The main thing to discuss is what are the wanted attributes and their uses in HA.

grea09 committed:
Addressing comments

async_update renamed back to update
revert removal of SCAN_INTERVAL = timedelta(seconds=120)
remove unecessary update during init.

grea09 committed:
reverting to "update"

grea09 committed:
not async

grea09 commented:
In that case I propose to make current_ban and total_ban exclusive. This way total_ban should be renamed past_ban with a list of unbaned ip. In that case the limit makes more sense. Otherwise total_ban will contain a mix of banned and unbanned ips. At least that's my view on it.

grea09 committed:
Proposed logic : separating banned and unbanned

I also changed the attributes names to reflect the change in the underlying logic. I know it will be even more breaking of a change but that will actually make sure people will not have silently wrong values in their install without warning.

I can easily revert this if asked.

grea09 committed:
Reflecting logic changes

grea09 commented:
Now it does with the updated logic. There's three attributes : `last_ban`, `banned_ips` and `unbanned_ips`. This way there's no more confusion on what's inside. I can also add the option of the limit if someone wants it. I know that changing names will make the breaking change worse but at least people won't just have to guess why their values are off and new users will understand what is in the attributes more easily.

Also, to add to the discussion: the reason I changed the value to number of ban was because it was more meaningful to have the graph of the evolution of the amount of bans instead of a random ip.

grea09 committed:
Needing back the timedelta dependancy

balloob commented on homeassistant/components/sensor/fail2ban.py:
```diff
@@ -27,9 +26,11 @@
 DEFAULT_NAME = 'fail2ban'
 DEFAULT_LOG = '/var/log/fail2ban.log'
 SCAN_INTERVAL = timedelta(seconds=120)
+DEFAULT_UNITS = 'ip'
```
This is not a unit, please remove.

grea09 replied:
I agree, that is not a unit. The problem is that the value is numerical and unless a unit is specified, the graphic log will treat the value as text and cause heavy persormance issue when displaying it. I have yet to find a workaround and I am willing to remove this ugly hack as soon as this is fixed.

balloob commented:
I agree that this PR should be split into 1 PR per feature. It is incredible difficult to see what has changed and for what reason.

grea09 commented:
I'll wait until the bug fixes are merged to submit the behaviour change as a PR.

grea09 committed:
Merge branch 'dev' into patch-4

grea09 commented:
Well, I let that PR open and merge the changes from the timer and ipv6 fixes.

balloob commented:
Since the features have been broken up and put into other PRs, I'll close this PR.

grea09 commented:
There was some features that were specific to this PR, namely the breaking change part. With this change the sensor extracts more useful informations. Should I do another PR or will this change be refused permanently ?
